### PR TITLE
gh-104504: Run mypy on cases_generator in CI (and blacken the code)

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-104504

### CODE REVIEW

**Comment by corona10** [comment]
I am not typing expert, so there might be a more promising way to attach a type hint for this.

- Add type hints where needed
- Add static symbols(OVERRIDE, REGISTER, IF) where needed
- Found a bug!

File: Tools/cases_generator/generate_cases.py
Comment (corona10) [inline Tools/cases_generator/generate_cases.py:324]: It was bug

File: Tools/cases_generator/analysis.py
Comment (corona10) [inline Tools/cases_generator/analysis.py]: FYI

````
analysis.py:175: error: Statement is unreachable  [unreachable]
                        typing.assert_never(thing)
````

**Comment by corona10** [comment]
cc @sobolevn

File: Tools/cases_generator/analysis.py
Comment (AlexWaygood) [reply Tools/cases_generator/analysis.py]: Huh, I thought that was the whole point of `assert_never` -- to have the type checker check that a branch of code you want to be unreachable really is unreachable. That seems like a bug in mypy if it's flagging these intentionally unreachable branches as unreachable :/

File: Tools/cases_generator/requirements-dev.txt
Comment (AlexWaygood) [inline Tools/cases_generator/requirements-dev.txt]:
```suggestion
# Requirements file for external linters and checks we run on Tools/cases_generator/ in CI
```

File: Tools/cases_generator/analysis.py
Comment (corona10) [reply Tools/cases_generator/analysis.py]: Even mypy 1.5.1 has the same issue. I think that it's worth reporting to the mypy team.

File: Tools/cases_generator/mypy.ini
Comment (AlexWaygood) [inline Tools/cases_generator/mypy.ini]: A _lot_ of the changes in this PR seem to be due to mypy flagging these intentionally unreachable `assert_never` branches as unreachable. Maybe we should disable this check for now and revert those changes in this PR, then consider in a separate PR whether it's worth it to enable this check?

```suggestion
```

File: Tools/cases_generator/requirements-dev.txt
Comment (AlexWaygood) [inline Tools/cases_generator/requirements-dev.txt]: Any reason not to use the latest version? :)

```suggestion
mypy==1.5.1
```

File: Tools/cases_generator/formatting.py
Comment (AlexWaygood) [inline Tools/cases_generator/formatting.py]: It would be better to use `collections.abc.Iterator` here, since `typing.Iterator` is deprecated

File: Tools/cases_generator/generate_cases.py
Comment (AlexWaygood) [inline Tools/cases_generator/generate_cases.py]: Here too

File: .github/workflows/mypy.yml
Comment (hugovk) [inline .github/workflows/mypy.yml]: Nitpick: let's put these jobs in alphabetical order, so this one first.

Or does it make sense to combine them into a single job, in their own steps? With a shared requirements file?

Comment (AlexWaygood) [reply .github/workflows/mypy.yml]: > Or does it make sense to combine them into a single job, in their own steps?

I think probably best to keep them separate. We might need different configs for each project -- looks like `warn_unreachable` causes quite a few false positives on `cases_generator`, whereas it has none on `clinic`: https://github.com/python/cpython/pull/108090#discussion_r1297514135.

We might also plausibly want different mypy versions for the two projects -- you often need to tweak your code slightly when upgrading mypy, and we have different experts for the two projects, so it might be hard to upgrade the mypy version for the two projects simultaneously.

**Review by AlexWaygood — COMMENTED** [review]
You might also want to alter `.github/dependabot.yml`, so that you get automated updates for the mypy version pin in `Tools/cases_generator/requirements-dev.txt`. Currently it only checks to see if dependencies in `Tools/clinic/` need updating:

https://github.com/python/cpython/blob/75b3db8445188c2ad38cabc0021af694df0829b8/.github/dependabot.yml#L15-L21

File: Tools/cases_generator/lexer.py
Comment (gvanrossum) [inline Tools/cases_generator/lexer.py]: I think we should just do all.

File: Tools/cases_generator/generate_cases.py
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py]: Also consider running black on all files here.

File: .github/workflows/mypy.yml
Comment (gvanrossum) [reply .github/workflows/mypy.yml]: OTOH there's quite a bit of duplication between the two sections. If something has to change we'd have to change it in each section. I don't how the first thing about how GitHub workflow .yml files work, so maybe that's unavoidable, but (as the expert for cases_generator and also a past mypy maintainer) I'd be happy to fix things quickly when a mypy upgrade requires it, and I'm sure we could coordinate with the AC expert.

File: Tools/cases_generator/analysis.py
Comment (gvanrossum) [reply Tools/cases_generator/analysis.py]: Not sure where my comment on this went, but I strongly prefer to keep those two lines (they exist specifically so a type checker can warn if you are missing a case). You could add
```
# type: ignore # mypy complains about this incorrectly
```
to the end of the line where the error occurs. Or just set `warn_unreachable = False` in `mypy.ini`.

File: Tools/cases_generator/generate_cases.py
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:26]: Good catch!

File: Tools/cases_generator/lexer.py
Comment (gvanrossum) [inline Tools/cases_generator/lexer.py:269]: FWIW, I can't comment on line 233, but it does some dynamic editing of `__all__`. That should just be replaced by a static list of all names exported. (This warning courtesy of pyright, the other type checker.)

**Review by gvanrossum — COMMENTED** [review]
Thanks so much for doing this!

File: .github/workflows/mypy.yml
Comment (AlexWaygood) [reply .github/workflows/mypy.yml]: Sure, I don't feel too strongly about it! We can merge the two sections and see how it goes. I think the AC expert here is me (at least when it comes to the type annotations in `Tools/clinic/`), so we'd just have to talk to each other :-)

If we want to have a common `requirements-dev.txt` file shared between the two projects, where should it live? Just in `Tools/`?

Comment (gvanrossum) [reply .github/workflows/mypy.yml]: > If we want to have a common `requirements-dev.txt` file shared between the two projects, where should it live? Just in `Tools/`?

I'd be fine with that, but I'm not really the expert on config files. :-)

Comment (corona10) [reply .github/workflows/mypy.yml]: I will create a common directory as Tools/.mypy it will be fine?

Comment (gvanrossum) [reply .github/workflows/mypy.yml]: Not a fan of `.` files everywhere (too easy to miss with `ls`).

File: .github/workflows/lint.yml
Comment (gvanrossum) [inline .github/workflows/lint.yml]: Sorry, that's not what I meant. I don't want black to run in CI. I just wanted you to run black after fixing those mypy issues. I'm not ready to run black on every commit.

**Comment by corona10** [comment]
@gvanrossum @AlexWaygood @hugovk 

Following things are updated
- Use static symbol exporting instead dynamic one.
- Replace deprecated typing.iterator into collections.abc one.
- Apply black and add a black step for case generator to Lint workflow.
- Use common Tools/requirements-dev.txt instead separate files
- Upgrade to the latest mypy

File: .github/workflows/mypy.yml
Comment (gvanrossum) [inline .github/workflows/mypy.yml]: Are there constraints on `x`? The cases generator uses `match`/`case` so it's got to be at least Python 3.10.

File: Tools/cases_generator/generate_cases.py
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py]: Those asserts need to stay -- per the comment on line 174-175, we're checking here that all targets of the pseudo have the same effects.

File: Tools/cases_generator/formatting.py
Comment (gvanrossum) [inline Tools/cases_generator/formatting.py]: I still like `typing.Iterator` better, it's so much shorter!

File: Tools/cases_generator/generate_cases.py
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:487]: Sorry, but I really want the `assert_never()` calls to stay (several times below also).

File: Tools/cases_generator/lexer.py
Comment (gvanrossum) [inline Tools/cases_generator/lexer.py:4]: Thanks for revamping lexer.py!

**Review by gvanrossum — COMMENTED** [review]
(Please don't `git push -f` -- it makes the GitHub review reset so I can't do "changes since last review".)

**Review by gvanrossum — COMMENTED** [review]
Thanks for taking black out.

File: Tools/cases_generator/formatting.py
Comment (corona10) [reply Tools/cases_generator/formatting.py]: @AlexWaygood Play rock-scissors-paper with Guido

File: Tools/cases_generator/generate_cases.py
Comment (corona10) [reply Tools/cases_generator/generate_cases.py:487]: Oh I missed some cases

File: .github/workflows/lint.yml
Comment (corona10) [reply .github/workflows/lint.yml]: Yeah, and there was some issue with CI so I rollbacked :)

File: .github/workflows/mypy.yml
Comment (corona10) [reply .github/workflows/mypy.yml]: Let's use 3.11. We need dog fooding

**Review by gvanrossum — APPROVED** [review]
LGTM. Don't worry about the origin of the Iterator type.

**Comment by corona10** [comment]
Waiting @AlexWaygood @hugovk comments before merging :)

File: .github/workflows/mypy.yml
Comment (erlend-aasland) [reply .github/workflows/mypy.yml]: I agree with Guido; `.` and `_` prefixed filenames are often hidden by various tools, which is irritating.

File: Tools/cases_generator/lexer.py
Comment (erlend-aasland) [inline Tools/cases_generator/lexer.py:1]: FYI: You could actually use Argument Clinic to generate the Python code for the keywords. (Argument Clinic supports Python files.) I'm not suggesting you do it for this PR, though; just a thought :)

**Review by erlend-aasland — APPROVED** [review]

File: Tools/cases_generator/formatting.py
Comment (AlexWaygood) [reply Tools/cases_generator/formatting.py]: Well, shorter yet is `from collections.abc import Iterator`!

File: Tools/cases_generator/mypy.ini
Comment (AlexWaygood) [inline Tools/cases_generator/mypy.ini]:
```suggestion
# Be strict:
```

File: Tools/cases_generator/mypy.ini
Comment (AlexWaygood) [inline Tools/cases_generator/mypy.ini:14]:
```suggestion

# Don't enable this one yet;
# it has a lot of false positives on `cases_generator`
warn_unreachable = False
```

File: Tools/cases_generator/formatting.py
Comment (corona10) [reply Tools/cases_generator/formatting.py]: > Well, shorter yet is from collections.abc import Iterator!

Applied!

File: Tools/cases_generator/lexer.py
Comment (corona10) [reply Tools/cases_generator/lexer.py:1]: Wow, Can I read the relevant documentation?

**Review by AlexWaygood — APPROVED** [review]
LGTM, thanks! :D

**Comment by gvanrossum** [comment]
Thanks!

File: .github/workflows/mypy.yml
Comment (hugovk) [reply .github/workflows/mypy.yml]: `3.x` is "the latest stable version of Python":

https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#specifying-a-python-version

That's 3.11 now, and will be 3.12 shortly after its release.

So `3.x` would be fine to minimise churn on this line (for next time).

Comment (corona10) [reply .github/workflows/mypy.yml]: > So 3.x would be fine to minimise churn on this line (for next time).

Good!

**Pull request merged**
